File: 3/3-3/Python/UDPReceiver.py
import socket
import time
import struct
import os


class UDPReceiver:
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    senderAddress = ('127.0.0.1', 9999)
    receiverAddress = ('127.0.0.1', 8888)
    s.bind(receiverAddress)

    seqPos = 0
    dataStartPos = 1
    crcLen = 2
    frameExpected = 0

    def getBinaryString(self, s):
        binaryStr = ""
        for i in range(len(s)):
            subStr = bin(ord(s[i])).replace('0b', '')
            while len(subStr) < 8:
                subStr = "0" + subStr
            binaryStr += subStr
        return binaryStr

    def getRemainderStr(self, dividendStr, divisorStr):
        dividendLen = len(dividendStr)
        divisorLen = len(divisorStr)
        dividend = int(dividendStr, 2)
        divisor = int(divisorStr, 2)
        dividend = dividend << (divisorLen - 1)
        divisor = divisor << (dividendLen - 1)
        sumLen = dividendLen + divisorLen - 1
        flag = pow(2, sumLen - 1)

        for i in range(dividendLen):
            if ((dividend & flag) != 0):
                dividend = dividend ^ divisor
                divisor = divisor >> 1
            else:
                divisor = divisor >> 1
            flag = flag >> 1

        remainderStr = bin(dividend).replace('0b', '')
        while len(remainderStr) < (divisorLen - 1):
            remainderStr = "0" + remainderStr
        return remainderStr

    def getCRCString(self, s):
        gxStr = "10001000000100001"
        return self.getRemainderStr(s, gxStr)

    def Receive(self):
        f = open("ReceiveText.txt", 'wb')
        while True:
            receiveFrame, address = self.s.recvfrom(1024)

            if str(receiveFrame) != "b'end'":

                binaryStr = self.getBinaryString(receiveFrame[self.dataStartPos:].decode('ISO-8859-1'))
                crcStr = self.getCRCString(binaryStr)

                if (int(crcStr, 2) == 0):
                    print("Current time: " + time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()))
                    print("frame_expected: %d" % self.frameExpected)
                    print("Data of the frame is right, serial is: %d" % receiveFrame[self.seqPos])

                    ack = 1 - self.frameExpected
                    sendFrame = struct.pack('B', ack)
                    self.s.sendto(sendFrame, self.senderAddress)
                    print("Sending ack, ack is: %d" % ack)
                    print()

                    if (receiveFrame[self.seqPos] == self.frameExpected):
                        self.frameExpected = (self.frameExpected + 1) % 2
                        f.write(receiveFrame[self.dataStartPos: 0 - self.crcLen])

                else:
                    print("Current time: " + time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()))
                    print("Data of the frame is wrong, doesn't send ack.")
                    print()

            else:
                print("Receive the file finished.")
                break

        f.close()
        self.s.close()


if __name__ == '__main__':
    print("Be ready to receive file...")
    operation = UDPReceiver()
    frameStr = operation.Receive()
    os.system("pause")

File: 3/3-3/Java/UDPReceiver.java
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketException;
import java.util.Arrays;
import java.text.SimpleDateFormat;
import java.util.Date;

public class UDPReceiver {
	private int senderPort = 9999;
	private int receiverPort = 8888;
	DatagramSocket ds;

	private int dataLen = 20;
	private int sendFrameLen = 1;
	private int receiveFrameLen = 25;
	private int serialPos = 0;
	private int dataStartPos = 1;
	private int crcStartPos = 21;
	private int isEndPos = 23;

	private int frameExpected = 0;

	public String getBinaryString(byte b[]) {
		String str = "";
		for (int i = 0; i < b.length; i++) {
			str += Integer.toBinaryString((b[i] & 0xFF) + 0x100).substring(1);
		}
		return str;
	}

	public String getRemainderStr(String dividendStr, String divisorStr) {
		int dividendLen = dividendStr.length();
		int divisorLen = divisorStr.length();
		for (int i = 0; i < divisorLen - 1; i++) {
			dividendStr += "0";
		}
		char str1[] = dividendStr.toCharArray();
		char str2[] = divisorStr.toCharArray();
		for (int i = 0; i < dividendLen; i++) {
			if (str1[i] == '1') {
				str1[i] = '0';
				for (int j = 1; j < divisorLen; j++) {
					if (str1[i + j] == str2[j]) {
						str1[i + j] = '0';
					} else {
						str1[i + j] = '1';
					}
				}
			}
		}
		String remainderStr = Arrays.toString(str1).replaceAll(", ", "").substring(dividendLen + 1,
				dividendLen + divisorLen);
		return remainderStr;
	}

	public String getCRCString(String s) {
		String gxStr = "10001000000100001";
		return getRemainderStr(s, gxStr);
	}

	public void Receive() throws Exception {
		try {
			ds = new DatagramSocket(receiverPort);
			InetAddress address = InetAddress.getByName(null);

			OutputStream os = new FileOutputStream(new File("ReceiveText.txt"));

			while (true) {
				byte[] receiveFrame = new byte[receiveFrameLen];
				DatagramPacket dp = new DatagramPacket(receiveFrame, receiveFrame.length);
				ds.receive(dp);

				if (receiveFrame[isEndPos] == 1) {
					System.out.println("Receive the file finished.");
					break;
				}
				
				int serial = receiveFrame[serialPos];

				int index = dataStartPos;
				for (; index < dataStartPos + dataLen; index++) {
					if (receiveFrame[index] == 0)
						break;
				}

				byte[] data = new byte[index - dataStartPos + 2];
				System.arraycopy(receiveFrame, dataStartPos, data, 0, index - dataStartPos);
				System.arraycopy(receiveFrame, crcStartPos, data, index - dataStartPos, 2);
				String binaryStr = getBinaryString(data);
				String crcStr = getCRCString(binaryStr);

				if (Integer.parseInt(crcStr, 2) == 0) {
					Date t = new Date();
					SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
					System.out.println("Current time: " + df.format(t));
					System.out.println("frame_expected: " + frameExpected);
					System.out.println("Data of the frame is right, serial is: " + serial);

					byte[] sendFrame = new byte[sendFrameLen];
					sendFrame[0] = (byte) (1 - frameExpected);
					DatagramPacket dp2 = new DatagramPacket(sendFrame, sendFrame.length, address, senderPort);
					ds.send(dp2);
					System.out.println("Sending ack, ack is: " + (1 - frameExpected));
					System.out.println();	
					
					if ((byte) frameExpected == serial) {
						frameExpected = (frameExpected + 1) % 2;
						os.write(data, 0, index - dataStartPos);
					}
				} else {
					Date t = new Date();
					SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
					System.out.println("Curremt time: " + df.format(t));
					System.out.println("Data of the frame is wrong, doesn't send ack.");
					System.out.println();
				}

			}

			os.close();
			ds.close();
		} catch (SocketException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}

	}

	public static void main(String[] args) throws Exception {
		System.out.println("Be ready to receive file...");
		UDPReceiver operation = new UDPReceiver();
		operation.Receive();
	}

}